core/src/main/java/ai/timefold/solver/core/config/solver/EnvironmentMode.java

```diff
@@ -106,22 +130,46 @@ public enum EnvironmentMode {
         this.asserted = asserted;
     }
 
+    public boolean isStepAssertOrMore() {
+        return switch (this) {
+            case NON_REPRODUCIBLE, NO_ASSERT, PHASE_ASSERT, STEP_ASSERT, NON_INTRUSIVE_FULL_ASSERT, FULL_ASSERT,
+                    TRACKED_FULL_ASSERT ->
+                isAsserted() && this != PHASE_ASSERT;
+            case REPRODUCIBLE -> PHASE_ASSERT.isAsserted();
+            case FAST_ASSERT -> STEP_ASSERT.isAsserted();
+        };
+    }
+
     public boolean isAsserted() {
         return asserted;
     }
 
     public boolean isNonIntrusiveFullAsserted() {
-        if (!isAsserted()) {
-            return false;
-        }
-        return this != FAST_ASSERT;
+        return switch (this) {
+            case NON_REPRODUCIBLE, NO_ASSERT, PHASE_ASSERT, STEP_ASSERT, NON_INTRUSIVE_FULL_ASSERT, FULL_ASSERT,
+                    TRACKED_FULL_ASSERT -> {
+                if (!isStepAssertOrMore()) {
+                    yield false;
+                }
+                yield this != STEP_ASSERT;
```

Comment:
Do we want to check if it is equal to `NON_INTRUSIVE_FULL_ASSERT`?

Reply:
Are `FULL_ASSERT` and `TRACKED_FULL_ASSERT` also considered non-intrusive full assertions?

Reply:
I have not changed the logic - whatever the logic was, it is still doing the same thing.